runtime/compiler/aarch64/codegen/ARM64PrivateLinkage.cpp

```diff
@@ -257,39 +645,42 @@ void TR::ARM64PrivateLinkage::createEpilogue(TR::Instruction *cursor)
-   TR::RealRegister::RegNum lastPreservedGPR = TR::RealRegister::x21;
-   for (TR::RealRegister::RegNum r = firstPreservedGPR; r >= lastPreservedGPR; r = (TR::RealRegister::RegNum)((uint32_t)r-1))
+   int32_t preservedRegisterOffsetFromJavaSP = cg()->getLargestOutgoingArgSize() + properties.getOffsetToFirstParm(); // outgoingArgsSize
+   TR::RealRegister::RegNum firstPreservedGPR = TR::RealRegister::x21;
+   TR::RealRegister::RegNum lastPreservedGPR = TR::RealRegister::x28;
+   for (TR::RealRegister::RegNum r = firstPreservedGPR; r <= lastPreservedGPR; r = (TR::RealRegister::RegNum)((uint32_t)r+1))
       {
       TR::RealRegister *rr = machine->getRealRegister(r);
       if (rr->getHasBeenAssignedInMethod())
          {
-         TR::MemoryReference *preservedRegMR = new (trHeapMemory()) TR::MemoryReference(javaSP, preservedRegisterOffset, cg());
+         TR::MemoryReference *preservedRegMR = new (trHeapMemory()) TR::MemoryReference(javaSP, preservedRegisterOffsetFromJavaSP, cg());
          cursor = generateTrg1MemInstruction(cg(), TR::InstOpCode::ldrimmx, lastNode, rr, preservedRegMR, cursor);
-         preservedRegisterOffset += 8;
+         preservedRegisterOffsetFromJavaSP += 8;
          }
       }
 
    // remove space for preserved registers
-   uint32_t frameSize = cg()->getFrameSizeInBytes();
-   if (constantIsUnsignedImm12(frameSize))
+   const TR::ARM64LinkageProperties& linkageProperties = getProperties();
+   int32_t firstLocalOffset = linkageProperties.getOffsetToFirstLocal();
+
+   uint32_t alignedFrameSizeIncludingReturnAddress = cg()->getFrameSizeInBytes() - firstLocalOffset;
+   if (constantIsUnsignedImm12(alignedFrameSizeIncludingReturnAddress))
       {
-      cursor = generateTrg1Src1ImmInstruction(cg(), TR::InstOpCode::addimmx, lastNode, javaSP, javaSP, frameSize, cursor);
+      cursor = generateTrg1Src1ImmInstruction(cg(), TR::InstOpCode::addimmx, lastNode, javaSP, javaSP, alignedFrameSizeIncludingReturnAddress, cursor);
       }
    else
       {
       TR::RealRegister *x9Reg = machine->getRealRegister(TR::RealRegister::RegNum::x9);
-      cursor = loadConstant32(cg(), lastNode, frameSize, x9Reg, cursor);
+      cursor = loadConstant32(cg(), lastNode, alignedFrameSizeIncludingReturnAddress, x9Reg, cursor);
       cursor = generateTrg1Src2Instruction(cg(), TR::InstOpCode::addx, lastNode, javaSP, javaSP, x9Reg, cursor);
       }
 
    // restore return address
-   TR::RealRegister *lr = machine->getRealRegister(TR::RealRegister::x30); // lr
+   TR::RealRegister *lr = machine->getRealRegister(TR::RealRegister::lr);
    if (machine->getLinkRegisterKilled())
       {
-      TR::MemoryReference *returnAddressMR = new (cg()->trHeapMemory()) TR::MemoryReference(javaSP, 0, cg());
-      cursor = generateTrg1MemInstruction(cg(), TR::InstOpCode::ldrimmx, lastNode, lr, returnAddressMR, cursor);
+      TR::MemoryReference *returnAddressMR = new (cg()->trHeapMemory()) TR::MemoryReference(javaSP, -8, cg());
```

Comment:
Perhaps you want to use `properties.getOffsetToFirstLocal()` instead of `-8` as the offset for the return address here, in the same way as you do in `createPrologue()`.